### PR TITLE
Support minimum stability for com_joomlaupdate

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Implement support for minimum stability for the core.

### Testing Instructions

- Switch to the testing update server (as of writing this it points to an RC update)
- check for updates
- `no downlaods found` message
- apply patch
- switch mimimum stablilty to RC in the updater config
- check for updates again.

### Expected result

Joomla offeres the RC version to update

### Actual result

image

### Documentation Changes Required

New option needs to be added to the com_joomlaupdate config docs page.

### Additional changes

- alpha order the language strings in com_joomlaupdate
- extend the list of possible issues for no download URL with minimum stability
- shorten the generation of $cache_timeout into one line.

### CODE REVIEW

File: administrator/components/com_joomlaupdate/models/default.php [hunk 1]
```diff
@@ -114,12 +114,17 @@ public function refreshUpdates($force = false)
...
-			$cache_timeout = 3600 * $cache_timeout;
+			$cache_timeout = 3600 * JComponentHelper::getParams('com_installer')->get('cachetimeout', 6, 'int');
 		}
 
-		$updater = JUpdater::getInstance();
+		$updater               = JUpdater::getInstance();
+		$minimumStability      = JUpdater::STABILITY_STABLE;
+		$comJoomlaupdateParams = JComponentHelper::getParams('com_joomlaupdate');
+
+		if (in_array($comJoomlaupdateParams->get('updatesource', 'nochange'), ['testing', 'custom']))
```
Comment (SharkyKZ) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
		if (in_array($comJoomlaupdateParams->get('updatesource', 'nochange'), array('testing', 'custom')))
```

Comment (Quy) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
		if (in_array($comJoomlaupdateParams->get('updatesource', 'nochange'), array('testing', 'custom')))
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 2]
```diff
@@ -172,10 +177,18 @@ public function getUpdateInformation()
 		$ret['latest']    = $updateObject->version;
 		$ret['hasUpdate'] = $updateObject->version != JVERSION;
 
+		$minimumStability      = JUpdater::STABILITY_STABLE;
+		$comJoomlaupdateParams = JComponentHelper::getParams('com_joomlaupdate');
+
+		if (in_array($comJoomlaupdateParams->get('updatesource', 'nochange'), ['testing', 'custom']))
```
Comment (SharkyKZ) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
		if (in_array($comJoomlaupdateParams->get('updatesource', 'nochange'), array('testing', 'custom')))
```

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 3]
```diff
@@ -33,37 +40,37 @@ COM_JOOMLAUPDATE_UPDATE_LOG_URL="Downloading update file from %s."
...
 COM_JOOMLAUPDATE_VIEW_DEFAULT_METHOD_FTP="Write files using FTP"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_METHOD_HYBRID="Hybrid (use FTP only if needed)"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_METHOD="Installation method"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_NOUPDATES="No updates available."
 COM_JOOMLAUPDATE_VIEW_DEFAULT_NOUPDATESNOTICE="You already have the latest Joomla version, %s."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_DOWNLOAD_URL="We can't find a download URL"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_DOWNLOAD_URL_DESC="An update to Joomla %1$s was found, but it wasn't possible to fetch the download URL for that update. There are three possibilities why this happens:<br>- Your host doesn't support <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a> and there is no alternative download for your configuration available.<br>- The update to Joomla %1$s is not available for your stability level.<br>- There is a problem with the Joomla Update Server.<br><br>Please try to download the update package from <a href="_QQ_"https://downloads.joomla.org/latest"_QQ_">the official Joomla download page</a> and use the Upload and Update tab."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_NO_DOWNLOAD_URL_DESC="An update to Joomla %1$s was found, but it wasn't possible to download that update. There are three possibilities why this happens:<br>- Your host doesn't support <a href="_QQ_"https://downloads.joomla.org/technical-requirements"_QQ_">the minimum requirements for Joomla %1$s</a> and there is no alternative download for your configuration available.<br>- The update to Joomla %1$s is not available for your stability level.<br>- There is a problem with the Joomla Update Server.<br><br>Please try to download the update package from <a href="_QQ_"https://downloads.joomla.org/latest"_QQ_">the official Joomla download page</a> and use the Upload and Update tab."
```

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 4]
```diff
@@ -33,37 +40,37 @@ COM_JOOMLAUPDATE_UPDATE_LOG_URL="Downloading update file from %s."
...
 COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_HOSTNAME="FTP Host"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_PASSWORD="FTP Password"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_PORT="FTP Port"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_FTP_USERNAME="FTP Username"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INFOURL="Additional Information"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALL_SELF_UPDATE_FIRST="You must update to the latest version of the Joomla Update Component before you can install the Joomla Core Update!"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLAGAIN="Reinstall Joomla core files"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLED="Installed Joomla version"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALLUPDATE="Install the Update"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALL_SELF_UPDATE_FIRST="You must update to the latest version of the Joomla Update Component before you can install the Joomla Core Update!"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_INSTALL_SELF_UPDATE_FIRST="You must update to the latest version of the Joomla Update Component before you can update Joomla!"
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 5]
```diff
@@ -128,11 +133,11 @@ public function refreshUpdates($force = false)
 		if (count($methodParameters) >= 4)
 		{
 			// Reinstall support is available in JUpdater
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php:135]: is this comment correct?

Comment (zero-24) [reply administrator/components/com_joomlaupdate/models/default.php:135]: why not? the latest com_joomlaupdate can be installed on accident jomla versions not supporting that feature.

Comment (brianteeman) [reply administrator/components/com_joomlaupdate/models/default.php:135]: it doesnt seem to be relevant to that code

Comment (zero-24) [reply administrator/components/com_joomlaupdate/models/default.php:135]: Can you please explain what you mean by that? The code mention checks whether there is support for the reinstall attribute or not and handle that case differently. In the line just after this comment it calls the new signature so to me this is still perfectly valid.

But I'm fine to change the text on your suggestion if required.

**Comment by twister65** [comment]
I have tested this item :white_check_mark: successfully on d9afae8febb73af3c8a65ede01dbc446fe46fc71<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25809">issues.joomla.org/tracker/joomla-cms/25809</a>.</sub>

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on d9afae8febb73af3c8a65ede01dbc446fe46fc71<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25809">issues.joomla.org/tracker/joomla-cms/25809</a>.</sub>

**Comment by ghost** [comment]
Status "Ready To Commit".

**Review by rdeutz — APPROVED** [review]

**Comment by zero-24** [comment]
Thanks for the merge and the tests 👍

**Pull request merged**